src/NuGet.Core/NuGet.Configuration/Settings/SettingsLoadingContext.cs

```diff
@@ -2,25 +2,45 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Globalization;
+using System.Collections.Concurrent;
 using System.IO;
-using System.Threading;
 using NuGet.Common;
 
 namespace NuGet.Configuration
 {
+    /// <summary>
+    /// Represents a cache context when loading <see cref="SettingsFile" /> objects so that they are only read once or if they change on disk.
+    /// </summary>
     public sealed class SettingsLoadingContext : IDisposable
     {
-        private readonly IList<Lazy<SettingsFile>> _settingsFiles = new List<Lazy<SettingsFile>>();
-        private readonly SemaphoreSlim _semaphore;
+        /// <summary>
+        /// A thread-safe cache for files based on their full path and last write time.
+        /// </summary>
+        private readonly ConcurrentDictionary<FileInfo, (DateTime LastWriteTime, Lazy<SettingsFile> Lazy)> _fileCache = new ConcurrentDictionary<FileInfo, (DateTime, Lazy<SettingsFile>)>(FileSystemInfoFullNameEqualityComparer.Instance);
```

Comment:
Given that this is a short lived cache, what is point of keeping the last write time?

Ideally within 1 operation, we used 1 version of the nuget.config and we don't care if it gets edited midway through as it's only likely to cause problems imo.

Reply:
I'd like to get more traction on this being used in Visual Studio so that the NuGet.config is only read once.  Long running processes would obviously benefit more from this logic but technically each solution restore operation could use a new context and not need to reload files that changed.  I don't feel strongly one way or the other, your call if I should simplify the cache to only use path instead of invalidating a cached item if the file changes.

Reply:
> I'd like to get more traction on this being used in Visual Studio so that the NuGet.config is only read once

I think when restore gets triggered the configs get read off disk + the VS Settings initialization so it happens only twice I think. 

@zivkan is working on a separate task for reloading configs in Visual Studio,https://github.com/NuGet/Home/issues/1538. 

I'm slightly concerned about the fact that there's a chance that 2 different versions of the same config are being used for different projects. 

Maybe @zivkan can share whether this would be useful for the config reload he's doing. I'm not sure this API is even consider as the VS config reload likely involves file watchers to trigger a reload.

Reply:
> I'd like to get more traction on this being used in Visual Studio so that the NuGet.config is only read once

VS already is only reading config files once. It just works in a different way. Basically all NuGet APIs take in an `ISettings` parameter, and all our VS code pass in the singleton `VsSettings` object, so it's only read once. The only current possible exception is if the customer edits the package sources in tools-options, although I'm not sure if that's implemented by writing the nuget.configs and then re-loading, or if it manages to updates settings in-memory without reloading. But as Nikolche mentioned, I'm working on detecting file changes, so reloading should become more common.

I know there is (or was?) a watson bug related to NuGet blocking threads (not async) while loading config, but given in the past we basically never re-loaded config, this PR wouldn't really help. It will reduce the duration of config reload once I implement that, but we don't yet have evidence that config load duration is a problem, so I can't shake the feeling that this might be solving a problem we don't have.

However, unlike restore, or VS, where we have a single-threaded entry point to load config once and then pass the instance around, I don't think the design of the MSBuild SDK Resolver allows us to do something similar.  We' had MULTIPLE bugs in the time I've been in the team related to the fact that the NuGetMSBuildSdkResolver gets called in parallel, but other parts of our initialization code expected initialization to happen single-threaded. So, I'm sure this config caching is important for the msbuild sdk resolver.

As mentioned/hinted in another comment, the problem with caches is you also need to know when to invalidate. If config loading isn't a measured perf issue, perhaps we'll avoid bugs if we avoid adding more caching layers. We already have a solution for restore and a different solution in VS, and this 3rd solution for the SDK resolver.  If the goal is to re-architect to try to have a single solution, rather than multiple, this PR is only a first step. But I'd want to understand the overall design, to know if it's a worthwhile step.

Reply:
Okay I'll update this PR to just cache the file based on path and fix the issue with it using a `Lazy<T>` which caches exceptions.